home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/VoteServiceImpl.java

```diff
@@ -0,0 +1,190 @@
+        "The number of voted questions does not equal the number of questions in the poll with 'id: %d'";
+
+    private static final String WRONG_QUESTIONS_FOR_POLL_MESSAGE =
+        "Some of the questions you are trying to vote do not match the poll with 'id: %d'";
+
+    private static final String WRONG_ANSWER_COUNT_VALIDATION_MESSAGE =
+        "Wrong count of selected answers to poll question with 'id: %d' (there is should be min 1, max %d)";
+
+    private static final String ANSWER_DOES_NOT_MATCH_QUESTION_VALIDATION_MESSAGE =
+        "The answer variant with 'id: %d' cannot be chosen when voting on the question with 'id: %d'";
+
+    private final VoteRepository voteRepository;
+
+    private final QuestionVoteRepository questionVoteRepository;
+
+    private final PollQuestionRepository questionRepository;
+
+    private final PollRepository pollRepository;
+
+    private final ServiceMapper mapper;
+
+    @Transactional
+    @Override
+    public ReadVoteDto createVote(Long pollId, User currentUser, CreateVoteDto voteDto) {
+        validatePollStatus(pollId);
+        validateReVoting(pollId, currentUser);
+        validateQuestionVotesCount(pollId, voteDto);
+        validatePollQuestionsMatching(pollId, voteDto);
+        validateAnswerCounts(voteDto);
+        validateAnswersMatching(voteDto);
+        var newVote = new Vote();
+        newVote.setPollId(pollId);
+        newVote.setUser(currentUser);
+        var savedVote = voteRepository.save(newVote);
+        List<QuestionVote> savedQuestionVotes = new ArrayList<>();
+        for (CreateQuestionVoteDto dto : voteDto.getQuestionVoteDtos()) {
+            QuestionVote newQuestionVote = mapper.convert(dto, QuestionVote.class);
+            newQuestionVote.setVote(savedVote);
+            var savedQuestionVote = questionVoteRepository.save(newQuestionVote);
+            savedQuestionVotes.add(savedQuestionVote);
+        }
+        savedVote.setQuestionVotes(savedQuestionVotes);
+        var reSavedVote = voteRepository.save(savedVote);
+        List<ReadQuestionVoteDto> questionVoteDtos = new ArrayList<>();
+        for (QuestionVote qv : reSavedVote.getQuestionVotes()) {
+            Long questionId = qv.getQuestionId();
+            PollQuestionType questionType =
+                questionRepository.findById(questionId).orElseThrow(() -> new NotFoundHomeException(
+                    String.format(NOT_FOUND_MESSAGE, "Poll question", questionId))).getType();
+            if (questionType.equals(PollQuestionType.ADVICE)) {
+                questionVoteDtos.add(mapper.convert(qv, ReadAdviceQuestionVoteDto.class));
+            } else {
+                questionVoteDtos.add(mapper.convert(qv, ReadMultipleChoiceQuestionVoteDto.class));
+            }
+        }
+        ReadVoteDto readVoteDto = new ReadVoteDto();
+        readVoteDto.setId(reSavedVote.getId());
+        readVoteDto.setQuestionVoteDtos(questionVoteDtos);
+        return readVoteDto;
```

Comment:
As I now we must have Vote and VoteDto, that VoteDto send to presentation layer for mapping. See how we are doing in another service.

Reply:
I was unable to do one DTO for both the Vote creation and the return to the presentation layer. So I made separate CreateVoteDto and ReadVoteDto.